src/ApolloHooksMutation.re

```diff
@@ -0,0 +1,210 @@
+type controlledResult('a) = {
+  loading: bool,
+  called: bool,
+  data: option('a),
+  error: option(error),
+};
+
+type controlledVariantResult('a) =
+  | Loading
+  | Called
+  | Data('a)
+  | Error(error)
+  | NoData;
+
+[@bs.module "graphql-tag"] external gql: ReasonApolloTypes.gql = "default";
+
+type mutationResult('a) = {. "data": 'a};
+
+[@bs.deriving abstract]
+type options('a) = {
+  [@bs.optional]
+  variables: Js.Json.t,
+  [@bs.optional]
+  client: ApolloClient.generatedApolloClient,
+  [@bs.optional]
+  refetchQueries,
+  [@bs.optional]
+  awaitRefetchQueries: bool,
+  [@bs.optional]
+  update: (ApolloClient.generatedApolloClient, mutationResult('a)) => unit,
+};
+
+type jsResult = {
+  .
+  "data": Js.Nullable.t(Js.Json.t),
+  "loading": bool,
+  "called": bool,
+  "error": Js.Nullable.t(error),
+};
+
+type jsMutate('a) = (. options('a)) => Js.Promise.t(jsResult);
+type mutation('a, 'b) =
+  (
+    ~mutation: 'b,
+    ~client: ApolloClient.generatedApolloClient=?,
+    ~refetchQueries: refetchQueries=?,
+    ~awaitRefetchQueries: bool=?,
+    unit
+  ) =>
+  Js.Promise.t(controlledVariantResult('a));
+
+[@bs.module "@apollo/react-hooks"]
+external useMutationJs:
+  (. ReasonApolloTypes.queryString, options('a)) => (jsMutate('a), jsResult) =
+  "useMutation";
+
+exception Error(string);
+
+let useMutation =
+    (
+      ~incompleteMutation=?,
+      ~mutation=?,
+      ~client=?,
+      ~refetchQueries=?,
+      ~awaitRefetchQueries=?,
+      ~update=?,
+      (),
+    ) => {
+  let (jsMutate, jsResult) =
+    useMutationJs(.
+      gql(.
+        switch (mutation, incompleteMutation) {
+        | (Some(mutation), _) => mutation##query
+        | (_, Some(incompleteMutation)) => incompleteMutation
+        | (None, None) =>
+          raise(
+            Error("You need to provide a mutation or an incomplete mutation"),
+          )
+        },
+      ),
+      options(
+        ~variables=?
+          switch (mutation, incompleteMutation) {
+          | (Some(mutation), _) => Some(mutation##variables)
+          | (_, Some(_incompleteMutation)) => None
+          | (None, None) =>
+            raise(
+              Error(
+                "You need to provide a mutation or an incomplete mutation",
+              ),
+            )
+          },
+        ~client?,
+        ~refetchQueries?,
+        ~awaitRefetchQueries?,
+        ~update?,
+        (),
+      ),
+    );
+
+  let parse = React.useRef(None);
+  let mutate =
+    React.useMemo1(
+      (
+        (),
+        ~mutation as passedMutation: 'b=?,
+        ~client=?,
+        ~refetchQueries=?,
+        ~awaitRefetchQueries=?,
+        (),
+      ) => {
+        let mutation =
+          switch (passedMutation, mutation) {
+          | (Some(mutation), _) => mutation
+          | (_, Some(mutation)) => mutation
+          | (None, None) => raise(Error("Need to pass a mutation"))
+          };
+
+        React.Ref.setCurrent(parse, Some(mutation##parse));
```

Comment:
Oh, this runs inside of `useMemo`? Callback passed into `useMemo` must be pure because it will run on every render, see [react docs](`https://reactjs.org/docs/hooks-reference.html#usememo`). And later it is called `jsMutate` which will send the actual query, is it correct?
 I am not sure it is a good idea to implement it this way, I think we will have to rethink this approach a bit, what do you think, @fakenickels ?

Reply:
Sorry, my bad, spent some time alazying the code and realised that it is memoizing the function and not actually calling it 🤦‍♀ 😄

Reply:
I merged this change from master, but good suggestion. useCallback is just a shortcut for memoizing functions with useMemo if I remember correctly.